### PR TITLE
Add neureka support

### PR DESCRIPTION
**Pull request by lukamac**

### CODE REVIEW

File: ne16/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    - [x] With
+    - [ ] Without
+    - [x] Relu (w/ and w/o)
+    - [x] Bias (w/ and w/o)
+    - [ ] Per-channel shift
+    - [x] Per-layer shift
+    - [ ] Rounding
+- [ ] Input type
+    - [x] uint8
+    - [ ] uint16
```
Comment (Scheremo) [inline ne16/README.md:23]: Shouldn't there also be int8?

File: ne16/gvsoc/ne16_gvsoc.h [hunk 2]
```diff
@@ -42,13 +42,17 @@ typedef enum ne16_gvsoc_log_level_e {
 static void ne16_gvsoc_log_activate(ne16_dev_t *dev,
                                     ne16_gvsoc_log_level_e log_level,
                                     ne16_gvsoc_log_format_e format) {
+#if __PLATFORM__ == ARCHI_PLATFORM_GVSOC
```
Comment (Scheremo) [inline ne16/gvsoc/ne16_gvsoc.h]: This depends on PULP SDK, right? It might be good to add comment in the README for people using this library without PULP SDK.

File: neureka/bsp/neureka_siracusa_bsp.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+void neureka_siracusa_cg_enable() {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR |=
+      NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_CG_EN;
+}
+
+void neureka_siracusa_cg_disable() {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR &=
+      ~NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_CG_EN;
+}
```
Comment (Scheremo) [inline neureka/bsp/neureka_siracusa_bsp.c]: There are no clock gates for N-EUREKA in Siracusa :)

File: neureka/hal/neureka.c [hunk 4]
```diff
@@ -18,15 +18,22 @@
...
-  unsupportedWeightOffsetMode,
-  unsupportedFeatureBitwidth,
-  dimensionMismatch
-} nnx_error_code;
+#define NEUREKA_STATUS_EMPTY (0x000)
+#define NEUREKA_STATUS_FULL (0x101)
 
-#endif // __NE16_ERROR_CODES_H__
\ No newline at end of file
+inline int neureka_task_queue_size(neureka_dev_t *dev) { return 2; }
```
Comment (Scheremo) [inline neureka/hal/neureka.c]: :)

File: neureka/hal/neureka_task.c [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "neureka_task.h"
+#include "neureka_task_defs.h"
+#include "pulp_nnx_util.h"
+
+inline uint32_t neureka_get_tile_padding(uint32_t padding, uint32_t i_height,
```
Comment (Scheremo) [inline neureka/hal/neureka_task.c]: NIT: Are you sure about force inling this? Looks like it's used in numerous places and it's a decent amount of code...

File: neureka/hal/neureka_task.c [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                                          : NEUREKA_FLAG_MODE_3x3;
+
+  task->data.cfg.conf0 |=
+      flag_input_signed << NEUREKA_SHIFT_FLAG_INPUT_SIGNED |
+      NEUREKA_FLAG_NORM_QUANT | quant.function | quant.mode |
+      (quant.shift_amount << 16) |
+      quant.flag_rounding << NEUREKA_SHIFT_ROUNDING | norm.mode |
+      norm.flag_bias << NEUREKA_SHIFT_FLAG_NORM_BIAS |
+      norm.flag_shift << NEUREKA_SHIFT_FLAG_NORM_SHIFT | NEUREKA_FLAG_USE_TCDM |
+      weights_offset_mode | flag_mode | (weights_bits - 1);
```
Comment (Scheremo) [inline neureka/hal/neureka_task.c]: I'd put a link to the register map of N-EUREKA here. While this looks correct, it seems hard to verify without this information.

File: neureka/hal/neureka_task.c [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+  const neureka_stride_t input_stride = {
+      .d0 = k_in_stride,
+      .d1 = k_in_stride * w_in_stride,
+      .d2 = 0 // Unused
+  };
+  task->data.cfg.input_stride = input_stride;
+
+  const neureka_stride_t output_stride = {
+      .d0 = 32, // TODO: should depend on outbytes. Probably 32 / outbytes
```
Comment (Scheremo) [inline neureka/hal/neureka_task.c]: I'd verify this TODO before merging

File: neureka/hal/neureka_task.c [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+  const uint16_t num_Wo = divnceil(w_out, NEUREKA_COMPUTE_SIZE_WIDTH);
+
+  const uint16_t rem_Ko = remainder(k_out, task->output_channel_throughput);
+  const uint16_t rem_Ki = remainder(k_in, task->input_channel_throughput);
+  const uint16_t rem_Ho = remainder(h_out, NEUREKA_COMPUTE_SIZE_HEIGHT);
+  const uint16_t rem_Wo = remainder(w_out, NEUREKA_COMPUTE_SIZE_WIDTH);
+  const uint16_t rem_Hi =
+      rem_Ho == 0 ? 0
+                  : (task->kernel_shape == 1 ? rem_Ho : rem_Ho + 2) -
+                        padding_bottom; // TODO: Check padding bottom
```
Comment (Scheremo) [inline neureka/hal/neureka_task.c]: SEMI-NIT: This is a bit hard to follow, it might be worth it to extract the second ternary operator to help readability.

File: test/TestClasses.py [hunk 9]
```diff
@@ -32,6 +32,10 @@ def implies(a: bool, b: bool):
     return (not a) or b
 
 
+def xor(a: bool, b: bool):
+    return (a and not b) or (not a and b)
```
Comment (Scheremo) [inline test/TestClasses.py]: Why not ^?

File: test/app/Makefile [hunk 10]
```diff
@@ -58,7 +60,10 @@ APP_SRCS += $(wildcard gen/src/*.c)
 
 # Flags
 
-APP_CFLAGS += $(INC_FLAGS) -O2 -w -Wall -Werror -flto
+ACCELERATOR_UPPERCASE := $(shell echo $(ACCELERATOR) | tr [:lower:] [:upper:])
+APP_CFLAGS += -DNNX_ACCELERATOR=\"$(ACCELERATOR)\" -DNNX_$(ACCELERATOR_UPPERCASE)
+
+APP_CFLAGS += -O2 -w -Wall -Werror -flto
 APP_LDFLAGS += -flto
```
Comment (Scheremo) [inline test/app/Makefile]: I would remove `-flto` (which can cause sporadic problems on the RISCV GCC Toolchain in PULP Platform) and specify the c standard `-std=cXX` where XX is the version you prefer.

File: neureka/bsp/neureka_siracusa_bsp.c (hunk 3 above)
Comment (FrancescoConti) [reply neureka/bsp/neureka_siracusa_bsp.c]: 👍  but we should still keep the stubs as that is definitely a bug, not a feature :D 
BTW isn't there the magic trick? Did you ever manage to make it work?

Comment (Scheremo) [reply neureka/bsp/neureka_siracusa_bsp.c]: About the stub: Yes! But I would take the opportunity to clarify that this won't make a functional difference.
The `magic trick` we discussed ~1 year ago? That one didn't really pull the power consumption down to the level I'd expect, and I didn't manage to reliably clear the resulting state :/ But I didn't go through RTL sim, so take it with a grain of salt.

**Comment by Scheremo** [comment]
Ciao Luka,
Great work, the PR is in very good shape, I would not label any of my comments as blocking.
I think it's a good idea to resolve the TODO in `neureka/hal/neureka_task.c`, but, feel free to resolve my other comments.

It would be good to add a version branch / commit hash for the PULP-SDK / GVSoC versions this was tested with, if possible, maybe in the README? On that note, we know about a bug in the pointwise mode in Siracusa, where only unsigned weights compute correctly. Is this captured in the GVSoC version you are using?

File: ne16/README.md (hunk 1 above)
Comment (lukamac) [reply ne16/README.md:23]: for NE16, no

File: neureka/bsp/neureka_siracusa_bsp.c (hunk 3 above)
Comment (lukamac) [reply neureka/bsp/neureka_siracusa_bsp.c]: Thanks for pointing this out! I prefer removing them if they are not working.

File: neureka/hal/neureka_task.c (hunk 5 above)
Comment (lukamac) [reply neureka/hal/neureka_task.c]: I will remove and let the compiler decide.

File: test/TestClasses.py (hunk 9 above)
Comment (lukamac) [reply test/TestClasses.py]: For some reason I thought python didn't have the xor operator...

File: test/app/Makefile (hunk 10 above)
Comment (lukamac) [reply test/app/Makefile]: I wanted it checked with `-flto` since I believe we use it often to get better performance. So to make sure it works, at least at kernel level, I added it. @Scheremo does it make sense?
Regarding the `-std=cXX` flag, I've never used that. Do you have any recommendations?

File: neureka/bsp/neureka_siracusa_bsp.c (hunk 3 above)
Comment (FrancescoConti) [reply neureka/bsp/neureka_siracusa_bsp.c]: I think it's much better to keep an empty stub than an #ifdef SIRACUSA_FIXED or something, but your call...

**Comment by lukamac** [comment]
> Ciao Luka,
> Great work, the PR is in very good shape, I would not label any of my comments as blocking.
I think it's a good idea to resolve the TODO in neureka/hal/neureka_task.c, but, feel free to resolve my other comments.

Will do. Thanks for the comments!

> It would be good to add a version branch / commit hash for the PULP-SDK / GVSoC versions this was tested with, if possible, maybe in the README?

Yes, good idea.

> On that note, we know about a bug in the pointwise mode in Siracusa, where only unsigned weights compute correctly. Is this captured in the GVSoC version you are using?

I think it is not because I didn't do any tests with unsigned weights. @Scheremo can you point me to a gvsoc/pulp-sdk version that exhibits that behaviour?

File: neureka/hal/neureka.c (hunk 4 above)
Comment (lukamac) [reply neureka/hal/neureka.c]: Yes?

File: neureka/bsp/neureka_siracusa_bsp.c [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+void neureka_siracusa_hci_reset_max_stall() {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR &=
+      ~NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_HCI_MAXSTALL;
+}
+
+// TODO: Check if needed for neureka
+void neureka_siracusa_hci_set_max_stall(uint32_t max_stall) {
+  *(volatile uint32_t *)NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_ADDR |=
+      max_stall & NEUREKA_SIRACUSA_CLUSTER_CTRL_HWPE_MASK_HCI_MAXSTALL;
+}
```
Comment (lukamac) [inline neureka/bsp/neureka_siracusa_bsp.c]: @arpansur @Scheremo Is this a thing in siracusa/does it do anything?

File: neureka/bsp/neureka_siracusa_bsp.c (hunk 3 above)
Comment (lukamac) [reply neureka/bsp/neureka_siracusa_bsp.c]: My idea is to have a different `neureka_<board>_bsp.c` file for each platform. If a future platform uses neureka and provides clock gating for it, it will be implemented.

File: neureka/hal/neureka.c (hunk 4 above)
Comment (Scheremo) [reply neureka/hal/neureka.c]: Is the task queue size something that derives from the neureka_dev_t struct? Is it something that might vary with future implementations? If not I think a `#define` might be a good solution worth considering.

File: test/app/Makefile (hunk 10 above)
Comment (Scheremo) [reply test/app/Makefile]: I think it's not unreasonable, however, flto is a program-level option, which means that the code calling into your library would have to be compiled with flto. I suppose for a test app this isn't an issue, but it might lead to people copying this Makefile into a context where flto causes issues.

Regarding the C std, I think C99 and C11 are reasonable targets, but if you really want you can also go newer. I'd suggest NOT targeting GNU extended versions (as gcc does by default!) since those can lead to tricky compliance issues with LLVM.

File: neureka/hal/neureka.c (hunk 4 above)
Comment (FrancescoConti) [reply neureka/hal/neureka.c]: I believe this is the reflection of a hardware parameter that can be 2, 2, or sometimes 2. Just kidding. It can also be 1 (to define at accelerator design time) or, in theory, up to 4 with small changes, but in practice 2 works better than 1 at little added cost and it is very hard to find a use case for >2.
So I'd go for the `#define` :D

File: neureka/bsp/neureka_siracusa_bsp.c (hunk 11 above)
Comment (Scheremo) [reply neureka/bsp/neureka_siracusa_bsp.c]: I am semi-sure this does do something, at least in the GVSoC port. As for Siracusa, I am not sure, I've not tried playing with it.

Comment (FrancescoConti) [reply neureka/bsp/neureka_siracusa_bsp.c]: In Siracusa: it does, although we probably do not touch this very often. It controls the QoS settings of the interconnect (see https://ieeexplore.ieee.org/document/10247945 and the figure below in particular).
On GVSOC: I am not sure it does something (I don't think it models the interconnect to a sufficient degree of accuracy to make a difference).
image

File: neureka/hal/neureka_task.c (hunk 7 above)
Comment (lukamac) [reply neureka/hal/neureka_task.c]: Todo verified. It is in fact not probably `32 / outbytes` but just 32, which is the output bandwidth in bytes. Added a define.

File: test/TestClasses.py (hunk 9 above)
Comment (lukamac) [reply test/TestClasses.py]: Removed here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/772fd951234d8f0d69454d8aaec726f018f8e3a6

File: test/app/Makefile (hunk 10 above)
Comment (lukamac) [reply test/app/Makefile]: I removed the `-flto` here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/8232527ca95633abcee3c59f4eb648bbac257a06
I added `-std=c11` here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/df3f5ddc9d1f503e7579194a696fd731a64bf1b6 but had to check if it was compiled for GAP_SDK because they use gcc's `asm` instead of `__asm__`.
I'm not the happiest with how I check if we are compiling for GAP_SDK, so if anyone has suggestions, feel free to give them.

File: neureka/bsp/neureka_siracusa_bsp.c (hunk 11 above)
Comment (lukamac) [reply neureka/bsp/neureka_siracusa_bsp.c]: Removed todos https://github.com/pulp-platform/pulp-nnx/pull/2/commits/8b37485b2596ca44ed40436a1392b35f235a487d

File: neureka/hal/neureka_task.c (hunk 8 above)
Comment (lukamac) [reply neureka/hal/neureka_task.c]: Removed the `rem_* == 0` here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/d9c45efe27e0918e620221a9b0839f8e5eb81d89
This is never true due to the actual implementation of the remainder function, which either gives remainder or divisor if the remainder is 0.

Comment (lukamac) [reply neureka/hal/neureka_task.c]: Renamed remainder here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/40657fd10da6f0998425e8917725f1a3a12f34a1

File: neureka/hal/neureka.c (hunk 4 above)
Comment (lukamac) [reply neureka/hal/neureka.c]: Changed here https://github.com/pulp-platform/pulp-nnx/pull/2/commits/1a4f873327ea3b7271e0a313a34ce3661bad5d21

File: test/app/Makefile (hunk 10 above)
Comment (lukamac) [reply test/app/Makefile]: Removed `-std=c11` in the end and will add it for library only when I add cmake support.

File: ne16/gvsoc/ne16_gvsoc.h (hunk 2 above)
Comment (lukamac) [reply ne16/gvsoc/ne16_gvsoc.h]: Removed them from here and added them in the tests since the tests are done with pulp-sdk and the library user might not. Commit https://github.com/pulp-platform/pulp-nnx/pull/2/commits/9145445c16ce982fd1215f305389ee0ad68a0fe0
